Fix simultaneous reconnection handling and exponential reconnection interval

jaunis opened the pull request:
Hello,

I ran into an issue in which a 'close' and an 'error' event are fired simultaneously. Thus, 2 re-connections  are launched at the same time, and the second removes the listeners positioned on the socket by the first one.
This causes errors raised by the newly created socket not to be caught.

In this pull request, I fix this issue by rejecting a reconnection request if another one is pending.

I also fix the exponential reconnection interval : it is now limited to five times the initial interval.

Regards

Commit:
Fix simultaneous reconnection handling

Commit:
Limit reconnection interval to five times the configured interval, instead of growing exponentially

andrew-aladev commented:
Please add a comment about Sigmoid function you used.

andrew-aladev commented:
I don't think `reconnect` is designed for ingnoring new reconnection attempt while reconnecting. I think it should add a new retry attempt in queue. We can ask @nucleardreamer to provide this info.

andrew-aladev commented:
Please use `retry.pending` or something similar. It looks like @nucleardreamer is designed `retry` object for such usage.

Commit:
- Add comment about Sigmoïd function
- Move attribute pendingReconnect to the retry object

jaunis commented:
@andrew-aladev It's done

nucleardreamer commented:
Thanks for the commit @jaunis, and sorry about the radio silence. I will be able to review this later today

Commit:
Reconnect indefinitely when a negative maxRetries is provided

Commit:
Re-emit start event when reconnecting

Commit:
Fix reconnection handling :
- timeout is given in seconds
- additional security to avoid launching several reconnections simultaneously
